### PR TITLE
Added LSM9DS1 (GYRO+ACCEL), LSM9DS1 (MAG) and BMI160 (GYRO+ACCEL). Fi…

### PR DESCRIPTION
**Pull request by FAC94**
We tried to make the PX4 compatible with the sensors LSM9DS1 and BMI160 sensors.

As the LSM9DS1 has two different CS pins for MAG and ACCEL+GYRO, we decided to keep them into separate files and treat them as two different devices.

The LSM9DS1 MAG driver seems to be working properly.
The LSM9DS1 ACCEL+GYRO driver works well when compiled with the LSM9DS1 MAG and MPU9250.
The BMI160 driver works well when compiled with the LSM9DS1 MAG's and MPU9250.

When we try to implement BMI160 and LSM9DS1 ACCEL+GYRO at the same time, it gives:
"offboard mission init: ERROR" and
"Not ready to fly: Sensors need inspection".

PS - The drivers' comments might not be correct, as this is a raw version of the final drivers.

### CODE REVIEW

**Comment by LorenzMeier** [comment]
@julianoes Would be great if you could review and provide architectural feedback.

File: cmake/configs/nuttx_px4fmu-v2_default.cmake [hunk 1]
```diff
@@ -17,32 +19,36 @@ set(config_module_list
...
 	drivers/mpu9250
+	drivers/lsm9ds1
+	drivers/lsm9ds1mag
+	drivers/bmi160
 	drivers/lsm303d
 	drivers/l3gd20
 	drivers/hmc5883
 	drivers/ms5611
-	drivers/mb12xx
+	#drivers/mb12xx
```
Comment (julianoes) [inline cmake/configs/nuttx_px4fmu-v2_default.cmake]: Did you have to disable these drivers because of flash space?

File: cmake/configs/nuttx_px4fmu-v2_default.cmake [hunk 2]
```diff
@@ -118,6 +124,9 @@ set(config_module_list
 	lib/geo_lookup
 	lib/conversion
 	lib/launchdetection
+	lib/terrain_estimation
+	lib/runway_takeoff
+	lib/tailsitter_recovery
```
Comment (julianoes) [inline cmake/configs/nuttx_px4fmu-v2_default.cmake]: I don't think these changes should be in this PR.

File: src/drivers/boards/px4fmu-v2/board_config.h [hunk 3]
```diff
@@ -112,8 +116,14 @@ __BEGIN_DECLS
 #define GPIO_SPI_CS_EXT2	(GPIO_OUTPUT|GPIO_PUSHPULL|GPIO_SPEED_50MHz|GPIO_OUTPUT_SET|GPIO_PORTC|GPIO_PIN15)
 #define GPIO_SPI_CS_EXT3	(GPIO_OUTPUT|GPIO_PUSHPULL|GPIO_SPEED_50MHz|GPIO_OUTPUT_SET|GPIO_PORTC|GPIO_PIN13)
 
+#define	GPIO_SPI_CS_MAG		GPIO_SPI_CS_BARO
```
Comment (julianoes) [inline src/drivers/boards/px4fmu-v2/board_config.h]: I don't see any space between ```define``` and ```GPIO...```, confusing.

File: src/drivers/boards/px4fmu-v2/board_config.h [hunk 4]
```diff
@@ -121,6 +131,9 @@ __BEGIN_DECLS
 #define PX4_SPIDEV_BARO		3
 #define PX4_SPIDEV_MPU		4
 #define PX4_SPIDEV_HMC		5
+#define	PX4_SPIDEV_MAG		6
+#define	PX4_SPIDEV_ACCEL_GYRO	7
+#define	PX4_SPIDEV_BMI		8
```
Comment (julianoes) [inline src/drivers/boards/px4fmu-v2/board_config.h]: same here, maybe that's just github...

File: src/drivers/boards/px4fmu-v2/board_config.h [hunk 5]
```diff
@@ -134,6 +147,9 @@ __BEGIN_DECLS
 #define PX4_SPIDEV_EXT_ACCEL_MAG	PX4_SPIDEV_EXT2
 #define PX4_SPIDEV_EXT_GYRO		PX4_SPIDEV_EXT3
 
+#define	PX4_SPIDEV_EXT_MAG		PX4_SPIDEV_EXT_BARO
+#define	PX4_SPIDEV_EXT_ACCEL_GYRO	PX4_SPIDEV_EXT_ACCEL_MAG
```
Comment (julianoes) [inline src/drivers/boards/px4fmu-v2/board_config.h]: This is all a bit confusing if you have to read it in some weeks/months time. Could you add some comments about what is what?

File: src/drivers/boards/px4fmu-v2/board_config.h [hunk 6]
```diff
@@ -209,6 +225,7 @@ __BEGIN_DECLS
 #define GPIO_TIM1_CH4OUT	GPIO_TIM1_CH4OUT_2
 #define GPIO_TIM4_CH2OUT	GPIO_TIM4_CH2OUT_2
 #define GPIO_TIM4_CH3OUT	GPIO_TIM4_CH3OUT_2
+#define DIRECT_PWM_OUTPUT_CHANNELS	6
```
Comment (julianoes) [inline src/drivers/boards/px4fmu-v2/board_config.h]: Why is this change in here?

File: src/drivers/boards/px4fmu-v2/px4fmu2_init.c [hunk 7]
```diff
@@ -212,19 +212,6 @@ static struct sdio_dev_s *sdio;
...
-/*{*/
-/*return 1;*/
-/*}*/
-/*#else*/
-/*__EXPORT int matherr(struct exception *e)*/
-/*{*/
-/*return 1;*/
-/*}*/
-/*#endif*/
-
```
Comment (julianoes) [inline src/drivers/boards/px4fmu-v2/px4fmu2_init.c]: Please keep this in separate commits. I don't see how this is related to the sensors.

**Comment by julianoes** [comment]
@FAC94 First of all, thanks for your contribution, appreciated! Let's try to get this merged in...

I tried to review the pull request and made a few comments above. However, github doesn't let me comment on the main [lsm9ds1mag.cpp](https://github.com/FAC94/Firmware/blob/2f8f7c21af52c6b4c67ddc6d8b2c5cc20d7f95f6/src/drivers/lsm9ds1mag/lsm9ds1mag.cpp) file.

I propose you do the following:
- Clean up the strange changes which have snuck in (see above).
- Make sure lsm9ds1mag.cpp adheres to the coding style (e.g. tabs vs. spaces). Easiest is just to use Tools/fix_code_style.sh.
- Make a new pull request against master and hopefully it'll show the changes and I can give feedback on the driver.

**Comment by FAC94** [comment]
@julianoes Thanks for your reply and sorry for the mess on the code :)

I think maybe the pull request went wrong because some of the changes on some files were not made... (it was my first pull request so I might did some mistakes on some steps).

I will try to make some comments to explain my decisions and make a cleaner code, easier to understand. Maybe in a week I'll make a new pull request with that changes.

Thanks for your tips :)

**Comment by julianoes** [comment]
@FAC94 What's the status here?

**Comment by FAC94** [comment]
@julianoes Hello, sorry for not replying back in a while but I was busy with another project these days. I could figure that the reason of the problem was the BMI160, so probably is some configurations of the sensor that are not being done well.
I will try to make a new pull request with the updated files today.
